src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs

```diff
@@ -1603,14 +1783,75 @@ private static void VerifyDerivedTypeConstraints(IEdmModel model, IEdmVocabulary
         {
-            Debug.Assert(identifier != null && identifier.Length >= 1 && identifier[0] == ':');
-            Debug.Assert(bindingType != null);
+            qualifiedName = null;
+            parenthesisExpression = null;
+            anotherEscapeFunctionStarts = false;
+
+            if (bindingType == null)  // escape function is only for bound functions
+            {
+                return false;
+            }
+
+            IEdmFunction function = model.FindBoundOperations(bindingType).OfType<IEdmFunction>().FirstOrDefault(f => IsUrlEscapeFunction(model, f));
+
+            if (function == null)
+            {
+                return false;
+            }
+
+            string nextPiece = null;
+            StringBuilder sb = new StringBuilder();
+
+            while (this.TryGetNextSegmentText(out nextPiece))
+            {
+                if (sb.Length >= 1)
+                {
+                    sb.Append('/');
+                }
+
+                sb.Append(nextPiece);
+
+                if (nextPiece[nextPiece.Length - 1] == ':')
+                {
+                    break;
+                }
+            }
+
+            string identifier = sb.ToString();
+
+            if (identifier != null && identifier.Length >= 2 && identifier[identifier.Length - 2] == ':')
+            {
+                anotherEscapeFunctionStarts = true;
+            }
+
+            bool isComposableRequired = identifier.Length >= 1 && identifier[identifier.Length - 1] == ':';
+
+            if ((function.Parameters.FirstOrDefault().Type.Definition != bindingType) || (isComposableRequired && !function.IsComposable) || (!isComposableRequired && function.IsComposable))
+            {
+                IEdmFunction candidate = model.FindBoundOperations(bindingType).OfType<IEdmFunction>().FirstOrDefault(f => f.Parameters.FirstOrDefault().Type.Definition == bindingType && f.IsComposable == isComposableRequired && IsUrlEscapeFunction(model, f));
+                // if exact match found use that.
+                if (candidate != null)
+                {
+                    function = candidate;
+                }
+                else
+                {
+                    // Look for another function if composability does not match. 
+                    if (function.IsComposable != isComposableRequired)
+                    {
+                        function = null;
+                    }
+                }
+            }
+
+            // If no exact match found for binding type, use base type to find bound functions. 
+            if (function == null)
+            {
+                function = model.FindBoundOperations(bindingType).OfType<IEdmFunction>().FirstOrDefault(f => f.IsComposable == isComposableRequired && IsUrlEscapeFunction(model, f));
+            }

```

Comment:
Yikes! This is the third time we've called a fairly expensive FindBoundOperations and run a fairly expensive (and similar) LINQ query over the results. Any way we can optimize this? Maybe first determine the list of candidates:
   List<IEdmFunction> candidates = model.FindBoundOperations(bindingType).OfType<IEdmFunction>().Where(f=>IsUrlEscapeFunction(model, f)).ToList();
and then pick the best from that (presumably rather small) list?
   -exact match of bindingType and composability
   -any (base) type and composability
   -return false? (I think we decided composability had to exactly match?)

Reply:
slightly different but did what we talked about. 

---
In reply to: [369837925](https://github.com/OData/odata.net/pull/1621#discussion_r369837925) [](ancestors = 369837925)